### ISSUE TITLE
Look into upgrading to Django 4.x and Python 3.10

### ISSUE THREAD

**Issue by rfrenchseti** [SETI/rms-opus#1084]
https://docs.djangoproject.com/en/3.0/releases/3.0/

**Comment by juzen2003** [comment]
- Known problems of upgrading to either Django 3.1.2 or 3.0.10
  - Version 3.1.2: (Need to wait for update on sql_flush call in django_nose for python manage.py test to work)
    - python manage.py test can't be run. We need to wait for the update on sql_flush call in django_nose.
    - Reference: wemake-services/django-test-migrations#117
    - Workaround, modify line 405 in runner.py (under site-pacakges/django_nose) by changing from:
```flush_statements = connection.ops.sql_flush(style, tables, connection.introspection.sequence_list())```
to:
```flush_statements = connection.ops.sql_flush(style, tables)```
  - Version 3.0.10: (Need to add a workaround to remove session from cache in base.py to make it work)
    - To have python manage.py runserver to work, we need to add a workaround to remove session from cahce under python site-packages "django/contrib/sessions/backends/base.py", we don't need this workaround in 3.1.x. After workaround, runserver will work, and test will pass.
    - Reference: https://code.djangoproject.com/ticket/31592
    - Workaround, add these in base.py:
       ```
       from django.contrib.sessions.models import Session
       Session.objects.all().delete()
       ```
- Requirement for Django version 3.x:
  - Change "render_to_response" to "render"
    - Reference: https://stackoverflow.com/questions/55911435/django-importerror-cannot-import-name-render-to-response-from-django-shortcu
  - Dependencies:
    - djangorestframework==3.11.1
    - mysqlclient==2.0.1

**Comment by rfrenchseti** [comment]
Dave says:

I think it's ok for us to stay on Django 2.2 since it's still releasing new update (the latest 2.2.17 is expected to release on Nov). And at the same time we can wait for django_nose to release its new support on 3.1.x for the sql_flush, once it supports 3.1.x, we can upgrade to Django 3.1.x right away.

Note: To make django_nose sql_flush call work with Django 3.1.x, we can modify line 405 in runner.py (under site-pacakges/django_nose) by changing from:
```flush_statements = connection.ops.sql_flush(style, tables, connection.introspection.sequence_list())```
to:
```flush_statements = connection.ops.sql_flush(style, tables)```
With above modification, python manage.py test will run and pass, runserver will work fine too.

**Comment by rfrenchseti** [comment]
Deferring until django_nose is updated.

**Comment by rfrenchseti** [comment]
This is actually for Django 4.x now, as 3.x is starting its end-of-life journey soon.

**Comment by dstopp** [comment]
Note that nose is not supported in Python 3.10 and should be replaced by pytest, see notes in this link: https://ipython.readthedocs.io/en/stable/whatsnew/version7.html

**Comment by rfrenchseti** [comment]
Upgrading to Django 4.x should be done at the same time as supporting Python 3.10. Currently OPUS doesn't work with 3.10 because some part of the Collections module was deprecated that was used by Django nose. Thus as part of fixing the tests for Django 4.x we should fix support for Python 3.10 as well.
